server/datastore/mysql/hosts.go

```diff
@@ -1648,11 +1652,58 @@ WHERE
 }
 
 func (ds *Datastore) UpdateOSVersions(ctx context.Context) error {
+	counts, err := ds.gatherOSCountsDB(ctx)
+	if err != nil {
+		return ctxerr.Wrapf(ctx, err, "update aggregated stats for os versions")
+	}
+
+	teams, err := ds.TeamsSummary(ctx)
+	if err != nil {
+		return ctxerr.Wrapf(ctx, err, "update aggregated stats for os versions")
+	}
+
+	countsByTeamId := make(map[uint]*json.RawMessage)
+	for _, c := range *counts {
+		countsByTeamId[c.TeamID] = c.OSVersions
+	}
+
+	for _, t := range teams {
+		if countsByTeamId[t.ID] == nil {
+			var raw json.RawMessage
+			raw, err = json.Marshal([]fleet.OSVersion{})
+			if err != nil {
+				return err
+			}
+			countsByTeamId[t.ID] = &raw
+		}
+	}
+
+	sql := `
+INSERT INTO aggregated_stats (id, type, json_value) 
+VALUES (?, ?, ?) 
+ON DUPLICATE KEY UPDATE
+	json_value = VALUES(json_value),
+	updated_at = CURRENT_TIMESTAMP
+`
+	for id, counts := range countsByTeamId {
```

Comment:
Please suggest any improvements to the sql that might eliminate the need for this insert loop or otherwise improve performance at scale

Reply:
You could construct a bulk insert instead of looping. See http://github.com/fleetdm/fleet/blob/master/server/datastore/mysql/hosts.go#L794-L794 for example.